### PR TITLE
refactor-frontend-flow

### PR DESCRIPTION
**Pull request by jablonnc**
# The Issue

## Lack Of Testing

Currently the only way to perform any kind of testing for our workflows is by running it manually on GitHub. This involves manually copying the associated workflows over into a given repository, opening up fake pull requests and then manually verify the output upon running the action. This takes up unnecessary developer hours and also can lead to brittle solutions that are hard to test.

## Overly Complex Workflows

The frontend application and utlity runtimes currently leverage a single workflow called [standard_spa_pr.yml](https://github.com/JupiterOne/.github/blob/main/.github/workflows/standard_spa_pr.yml). This workflows is actually 3 separate workflows consolidated into one:
- Application Runtime PR - Our application repos [currently leverage](https://github.com/JupiterOne/web-alerts/blob/825328759eaab29dbe0631b26d7b8259b349b757/.github/workflows/pull_request.yml#L8) this workflow when a PR is opened. While similar to the `Utility Runtime PR` flow, the [last step](https://github.com/JupiterOne/.github/blob/458227285c359295339836e719004fadc7f12610/.github/workflows/standard_spa_pr.yml#L552C4-L552C4) that involves trigger E2E tests in another repo would never get called for this type of entity.
- Utility Runtime PR - Our utility runtime repos [currently leverage](https://github.com/JupiterOne/web-query-results/blob/7d39b4e3742b9ce005de1b886727fee58e096886/.github/workflows/pull_request.yml#L8C8-L8C8) this workflow when a PR is opened. Once again, while there is overlap with the `Application Runtime PR` above, there are a number of jobs that would never get triggered for this type of entity. Those being the steps related to running E2E tests ([prepare](https://github.com/JupiterOne/.github/blob/458227285c359295339836e719004fadc7f12610/.github/workflows/standard_spa_pr.yml#L343C3-L343C3) and [running](https://github.com/JupiterOne/.github/blob/458227285c359295339836e719004fadc7f12610/.github/workflows/standard_spa_pr.yml#L408)).
- E2E Trigger Response - The most problematic of all, our application runtime repos also [leverage this flow](https://github.com/JupiterOne/web-home/blob/6087b391b2eb740349327eb9162fa0fd8a2a5e65/.github/workflows/pull_request.yml#L8)  for the purpose of providing an action that can be triggered externally by other repos to run E2E tests (ex. I change a runtime module, I now need to run tests in 3 other repos to validate those integration tests are still passing with the new module). This particular setup ends up turning off a majority of the flow (setting the following inputs to false: use_validate, use_chromatic, use_magic_url), just so that it can leverage the jobs related to running E2E tests.

By trying to leverage one workflow to handle these various flows, we clearly have an issue with a separation of concerns. We have unnecessarily created a very complex environment for our CICD builds that is hard to predict, test, and improve the performance of.

# The Solution

This PR introduces the following:
- `Composite Actions` - Breaks up the various jobs in our workflow into a series of reusable composite actions. This enables us to easily share these actions among multiple workflows and makes things easier to understand and reason about.
- `Test Solution` - Introduces [mock-github](https://github.com/kiegroup/mock-github) and [act-js](https://github.com/kiegroup/act-js). Using these two libraries together, we are able to write unit tests around all of our workflows and composite actions.
- `Better Documentation` - As shown in the [updated README](https://github.com/JupiterOne/.github/tree/refactor-frontend-flow), we have added a lot of new documentation around each of our new workflows and their associated composite actions. Additionally we've added diagrams for each of our flows to also aid in understanding how things work.
- `New Deploy Flow` - Currently our [deployment flow](https://github.com/JupiterOne/web-alerts/blob/main/.github/workflows/deploy.yml) is not encapsulated in a shared workflow. This new workflow has now been added, meaning that all the workflows for our runtimes applications and utilities will now use a unified workflow moving forward.

# Next Steps

## NPM Workflows

We currently have two workflows related to our NPM packages. These workflows will be updated to take the same approach used for our runtime flows:
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/npm_package_pr_npm.yaml
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/npm_package_release_npm.yaml

## Deprecates the following workflows

- https://github.com/JupiterOne/.github/blob/main/.github/workflows/standard_spa_pr.yml
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/standard_spa_pr_NPM.yml
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/e2e_trigger.yml
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/cortex_frontend.yml
- https://github.com/JupiterOne/.github/blob/main/.github/workflows/jetpack_frontend.yaml

Once all application runtimes and utility runtimes have been switched over to use the new workflows, these will be removed.

### CODE REVIEW

File: .github/.DS_Store
Comment (ethan-steinman-j1) [inline .github/.DS_Store:1]: We should drop this file and add it to .gitignore

File: .github/actions/frontend/runtime/e2e_prepare/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Frontend Runtime - E2E Prepare
+
+This [composite action](./action.yml) is responsible for executing [Artemis](https://github.com/JupiterOne/artemis) and obtaining the necessary information needed for logging in via the account/user created.
+
+## Inputs
+
+| Name                        | Type    | Default                      | Required  | Description                                               |
+| --------------------------- | ------- | ---------------------------- | --------- | --------------------------------------------------------- |
+| `e2e_artemis_config_path`   | String  | cypress/artemis-config.yaml  | False     | Used to determine the path to the artemis config file
+| `userCount`                 | String  |                              | True      | The number of tests that you want Cypress to run in parallel (obtained via `$(echo '${{ inputs.e2e_containers }}' | jq '. | length')`)
```
Comment (ethan-steinman-j1) [inline .github/actions/frontend/runtime/e2e_prepare/README.md:10]: This is listed as require here, but has a default value in the action code.

File: .github/actions/frontend/runtime/e2e_prepare/action.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+runs:
+  using: "composite"
+  steps:
+    - uses: actions/checkout@v3
+    - name: e2e_prepare_inputs
+      shell: bash
+      run: |
+        echo "e2e_artemis_config_path=${{ inputs.e2e_artemis_config_path }}"
+        echo "user_count=${{ inputs.user_count }}"
```
Comment (ethan-steinman-j1) [inline .github/actions/frontend/runtime/e2e_prepare/action.yml:29]: is this effectively a logging step to help with debugging?

File: .github/actions/frontend/runtime/e2e_prepare/test/action_test.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        shell: bash
+      - name: results_artemis_account_subdomain
+        run: echo ${{ steps.e2e_prepare.outputs.artemis_account_subdomain }}
+        shell: bash
+      - name: results_artemis_account_id
+        run: echo ${{ steps.e2e_prepare.outputs.artemis_account_id }}
+        shell: bash
+      - name: results_artemis_users
+        run: echo ${{ steps.e2e_prepare.outputs.artemis_users }}
+        shell: bash
```
Comment (ethan-steinman-j1) [inline .github/actions/frontend/runtime/e2e_prepare/test/action_test.yml:26]: are these echo step required to make the test work?  mainly just trying to get my head wrapped around how the code work with this question.

File: .github/actions/frontend/runtime/e2e_prepare/action.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    - name: setup_env
+      if: ${{ !env.TEST }}
```
Comment (ethan-steinman-j1) [inline .github/actions/frontend/runtime/e2e_prepare/action.yml:31]: I assume running the tests automatically sets this env variable?

File: .github/.DS_Store
Comment (jablonnc) [reply .github/.DS_Store:1]: Addressed

File: .github/actions/frontend/runtime/e2e_prepare/README.md (hunk 1 above)
Comment (jablonnc) [reply .github/actions/frontend/runtime/e2e_prepare/README.md:10]: Good call, this should be false. Updated.

File: .github/actions/frontend/runtime/e2e_prepare/test/action_test.yml (hunk 3 above)
Comment (ethan-steinman-j1) [reply .github/actions/frontend/runtime/e2e_prepare/test/action_test.yml:26]: the wrapper workflow to call the action makes sense.  It appears to me that we're validating the outputs twice.  Once in action.test.ts via expect statements, and again on lines 15-26 via echo statements.

File: .github/workflows/frontend_runtime_e2e_trigger_response.yml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        type: string
+      external_pr_author:
+        description: 'Used by the e2e_trigger to give builds in Cypress the correct author name associated with the owner of the PR'
+        type: string
+      external_pr_sha:
+        description: 'Used by the e2e_trigger to give builds in Cypress the correct SHA associated with the PR that triggered the flow'
+        type: string
+      external_pr_repo_name:
+        description: 'Used by the e2e_trigger to tag builds in Cypress with the appropriate repo name associated with the repo that triggered the flow'
+        type: string
```
Comment (ethan-steinman-j1) [inline .github/workflows/frontend_runtime_e2e_trigger_response.yml:32]: These are listed as required in the associated docs.

File: .github/workflows/frontend_runtime_utility_pr.yml [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+          e2e_auto: ${{ secrets.E2E_AUTO }}
+          repo_name: ${{ matrix.repos.repo.name }}
+          repo_spec: ${{ matrix.repos.repo.spec }}
+
+  # Unless e2e_pass_on_error is true, this will never get hit if an E2E test fails
+  e2e_status:
+    runs-on: ${{ (inputs.fallback_runner && 'ubuntu-latest') || fromJson('["jupiterone-dev", "arm64"]') }}
+    needs: [e2e_trigger_remote_tests]
+    steps:
+      - uses: actions/checkout@v2
```
Comment (ethan-steinman-j1) [inline .github/workflows/frontend_runtime_utility_pr.yml:124]: same as application_pr workflow, is this needed?

File: README.md [hunk 7]
```diff
@@ -1,4 +1,117 @@
 # .github
 
-For documentation on organization-wide GitHub workflows, please see
-[.github/workflows/README.md](https://github.com/JupiterOne/.github/blob/main/.github/workflows/README.md)
+## Workflows
+
+The following workflows are for the entire JupiterOne organization, intending to consolidate t `JupiterOne Organization-Wide GitHub`
+
+### Frontend Workflows
```
Comment (ethan-steinman-j1) [inline README.md:7]: Missing utility PR workflow here

File: README.md [hunk 8]
```diff
@@ -1,4 +1,117 @@
...
+
+### Backend Workflows
+- [publish_docker_ghcr](.github/workflows/docs/backend/publish_docker_ghcr.md)
+- [publish_integration_docker_image](.github/workflows/docs/backend/publish_integration_docker_image.md)
+
+## Composite Actions
+
+The workflows above take advantage of the composite actions listed below, leveraging composition to execute those jobs that pertain to the workflow. The actions below enable us to consolidate repeatable jobs into a single location. This makes our workflows DRY, easier to reason about and more stable.
+
+### Frontend Workflows
```
Comment (ethan-steinman-j1) [inline README.md:19]:
```suggestion
### Frontend Actions
```

File: .github/actions/frontend/runtime/migration_number/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# Frontend Runtime - Extract Migration Number
+
+This [composite action](./action.yml) is responsible for running the `validation` command in a frontend repo. Additionally it is responsible for running the `remote-types test` command if applicable to determine if any breaking changes were made.
```
[the remark this answers is not in the document]
Comment (jablonnc) [reply .github/actions/frontend/runtime/migration_number/README.md:3]: Good eye, this has been updated.

File: .github/workflows/frontend_runtime_e2e_trigger_response.yml (hunk 5 above)
Comment (jablonnc) [reply .github/workflows/frontend_runtime_e2e_trigger_response.yml:32]: As they are, these have been updated.

File: README.md (hunk 7 above)
Comment (jablonnc) [reply README.md:7]: This has now been added.

File: README.md (hunk 8 above)
Comment (jablonnc) [reply README.md:19]: Updated

File: .github/actions/frontend/runtime/e2e_prepare/action.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    value: ${{ steps.artemis_info.outputs.artemis_account_subdomain }}
+  artemis_account_id:
+    value: ${{ steps.artemis_info.outputs.artemis_account_id }}
+  artemis_users:
+    value: ${{ steps.artemis_info.outputs.artemis_users }}
+
+runs:
+  using: "composite"
+  steps:
+    - uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (jablonnc) [reply .github/actions/frontend/runtime/e2e_prepare/action.yml]: I've removed the checkout from the composite actions, only the job that's leveraging the action should need to do this.

**Review by ethan-steinman-j1 — DISMISSED** [review]

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,4 @@
+node_modules/
+repo
+*.log
+.DS_Store
```
Comment (Sivli-Embir) [inline .gitignore:4]: Thanks for adding this

File: package.json [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  "license": "UNLICENSED",
+  "scripts": {
+    "clean": "npx rimraf repo",
+    "test": "npm run clean && vitest --no-threads",
+    "typecheck": "tsc -b",
+    "lint": "eslint . --ext .ts",
+    "validate": "npm run lint && npm run typecheck && npm run test"
+  },
+  "config": {
+    "migration": 1
```
Comment (Sivli-Embir) [inline package.json:16]: Why add a migration number here?

File: tests/utils/mock_composite_step.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        step.run = mockStep?.mockWith || `echo ""`;
+
+        if (step.uses) {
+          delete step['uses'];
+        }
+      }
+    });
+  });
+  
+  // console.log(newRepoAction?.runs?.steps);
```
Comment (Sivli-Embir) [inline tests/utils/mock_composite_step.ts:47]: dead code

File: package.json (hunk 12 above)
Comment (jablonnc) [reply package.json:16]: This is actually leveraged by the tests, you can see it [being copied over here](https://github.com/JupiterOne/.github/blob/7ae7f35e9983c31f9bd0f0c3e28d6402525e2902/tests/utils/setup.ts#L28) in the virtual repo that gets creatd.

File: tests/utils/mock_composite_step.ts (hunk 13 above)
Comment (jablonnc) [reply tests/utils/mock_composite_step.ts]: addressed

**Review by Sivli-Embir — APPROVED** [review]

**Review by electricgull — APPROVED** [review]

**Pull request merged**